polyglot/piranha/visualize_rules_graph.py

```diff
@@ -0,0 +1,234 @@
+import argparse
+import os
+import re
+import toml
+import graphviz
+from dataclasses import dataclass
+
+# REQUIRED: pip install toml
+# REQUIRED: pip install graphviz
+
+
+@dataclass(frozen=True, eq=True)
+class Edge:
+    to: str
+    scope: str
+
+
+def collect_rules_groups_edges():
+    for config_path in args.configurations_path:
+        config_path = os.path.abspath(config_path) + os.path.sep
+        rules_file = config_path + 'rules.toml'
+
+        # rules.toml should exist, deliberately fail otherwise
+        rules_toml_dict = toml.load(rules_file)
+        collect_rules_and_groups(rules_toml_dict)
+
+        # edges are optional for input rules (i.e., non built-in)
+        edges_file = config_path + 'edges.toml'
+        if os.path.exists(edges_file):
+            edges_toml_dict = toml.load(edges_file)
+            collect_edges(edges_toml_dict)
+
+
+def collect_rules_and_groups(rules_toml_dict):
+    """
+    Collects rules and groups to further build graph nodes.
+
+    Rules belonging to a group different than `Cleanup Rule` are displayed below the group name.
+
+    Cleanup Rule's are added to a set as they are displayed differently.
+        They get an additional label `(Cleanup_Rule)`.
+
+    Nodes without a group are added to a different set.
+    """
+    for rule_toml in rules_toml_dict['rules']:
+        rule_name: str = sanitize_name(rule_toml['name'])
+        if 'groups' in rule_toml:
+            rule_groups = rule_toml['groups']
+            collect_node_for_group_rule(rule_name, rule_groups)
+        else:  # rule without group
```

Comment:
why would node without a `group` be orphan node? Shouldn't it be when it is a node without an edge?